src/qunitmulti.cpp

```diff
@@ -0,0 +1,215 @@
+//////////////////////////////////////////////////////////////////////////////////////
+//
+// (C) Daniel Strano and the Qrack contributors 2017, 2018. All rights reserved.
+//
+// QUnitMulti is a multiprocessor variant of QUnit.
+// QUnit maintains explicit separability of qubits as an optimization on a QEngine.
+// See https://arxiv.org/abs/1710.05867
+// (The makers of Qrack have no affiliation with the authors of that paper.)
+//
+// Licensed under the GNU Lesser General Public License V3.
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/lgpl-3.0.en.html
+// for details.
+
+#include "qunitmulti.hpp"
+#include "common/oclengine.hpp"
+#include "qfactory.hpp"
+
+namespace Qrack {
+
+QUnitMulti::QUnitMulti(bitLenInt qBitCount, bitCapInt initState, qrack_rand_gen_ptr rgp, complex phaseFac, bool doNorm,
+    bool randomGlobalPhase, bool useHostMem, int deviceID, bool useHardwareRNG, bool useSparseStateVec,
+    real1 norm_thresh, std::vector<bitLenInt> devList)
+    : QUnit(QINTERFACE_OPENCL, qBitCount, initState, rgp, phaseFac, doNorm, randomGlobalPhase, useHostMem, -1,
+          useHardwareRNG, useSparseStateVec, norm_thresh)
+{
+    // Notice that this constructor does not take an engine type parameter, and it always passes QINTERFACE_OPENCL to
+    // the QUnit constructor. For QUnitMulti, the "shard" engines are therefore guaranteed to always be QEngineOCL
+    // types, and it's safe to assume that they can be cast from QInterfacePtr types to QEngineOCLPtr types in this
+    // class.
+
+    std::vector<DeviceContextPtr> deviceContext = OCLEngine::Instance()->GetDeviceContextPtrVector();
+
+    if (devList.size() == 0) {
+        defaultDeviceID = OCLEngine::Instance()->GetDefaultDeviceID();
+
+        for (bitLenInt i = 0; i < deviceContext.size(); i++) {
+            DeviceInfo deviceInfo;
+            deviceInfo.id = i;
+            deviceList.push_back(deviceInfo);
+        }
+
+        std::swap(deviceList[0], deviceList[defaultDeviceID]);
+    } else {
+        defaultDeviceID = devList[0];
+
+        for (bitLenInt i = 0; i < devList.size(); i++) {
+            DeviceInfo deviceInfo;
+            deviceInfo.id = devList[i];
+            deviceList.push_back(deviceInfo);
+        }
+    }
+
+    for (bitLenInt i = 0; i < deviceList.size(); i++) {
+        deviceList[i].maxSize = deviceContext[deviceList[i].id]->device.getInfo<CL_DEVICE_MAX_MEM_ALLOC_SIZE>();
+    }
+
+    if (devList.size() == 0) {
+        std::sort(deviceList.begin() + 1, deviceList.end(), std::greater<DeviceInfo>());
```

Comment:
If you're just doing sorts, then I think you only need the operator<.

Reply:
One of the `>` operators can be removed.